src/peer_channels.rs

```diff
@@ -0,0 +1,597 @@
+//!    This message contains the users relay-server and public key.
+//!    Direct IP address is not included as this information can be persisted by email providers.
+//! 4. After the announcement, the sending peer joins the gossip swarm with an empty list of peer IDs (as they don't know anyone yet).
+//! 5. Upon receiving an announcement message, other peers store the sender's [NodeAddr] in the database
+//!    (scoped per WebXDC app instance/message-id). The other peers can then join the gossip with `joinRealtimeChannel().setListener()`
+//!    and `joinRealtimeChannel().send()` just like the other peers.
+
+use anyhow::{anyhow, Context as _, Result};
+use iroh_gossip::net::{Gossip, JoinTopicFut, GOSSIP_ALPN};
+use iroh_gossip::proto::{Event as IrohEvent, TopicId};
+use iroh_net::magic_endpoint::accept_conn;
+use iroh_net::relay::{RelayMap, RelayUrl};
+use iroh_net::{key::SecretKey, relay::RelayMode, MagicEndpoint};
+use iroh_net::{NodeAddr, NodeId};
+use url::Url;
+
+use crate::chat::send_msg;
+use crate::config::Config;
+use crate::context::Context;
+use crate::message::{Message, MsgId, Viewtype};
+use crate::mimeparser::SystemMessage;
+use crate::EventType;
+
+impl Context {
+    /// Create magic endpoint and gossip for the context.
+    pub(crate) async fn init_peer_channels(&self) -> Result<()> {
+        let secret_key: SecretKey = self.get_or_generate_iroh_keypair().await?;
+        info!(self, "Secret key: {}", secret_key.public().to_string());
+
+        let mut ctx_gossip = self.iroh_gossip.write().await;
+        if ctx_gossip.is_some() {
+            warn!(
+                self,
+                "Tried to create endpoint even though there already is one"
+            );
+            return Ok(());
+        }
+
+        let endpoint = MagicEndpoint::builder()
+            .secret_key(secret_key)
+            .alpns(vec![GOSSIP_ALPN.to_vec()])
+            .relay_mode(RelayMode::Custom(RelayMap::from_url(
+                self.metadata
+                    .read()
+                    .await
+                    .as_ref()
+                    .map(|conf| {
+                        let url = conf.iroh_relay.as_deref().context("can't get relay url")?;
+                        Ok::<_, anyhow::Error>(RelayUrl::from(Url::parse(url)?))
+                    })
+                    .transpose()?
+                    .unwrap_or(RelayUrl::from(
+                        Url::parse("https://iroh.testrun.org:4443").unwrap(),
+                    )),
+            )))
+            .bind(0)
+            .await?;
+
+        // create gossip
+        let my_addr = endpoint.my_addr().await?;
+        let gossip = Gossip::from_endpoint(endpoint.clone(), Default::default(), &my_addr.info);
+
+        // spawn endpoint loop that forwards incoming connections to the gossiper
+        let context = self.clone();
+        tokio::spawn(endpoint_loop(context, endpoint.clone(), gossip.clone()));
```

Comment:
this task is not managed at all (shutdown etc)

Reply:
We want this to be shut down with delta chat, so this should be fine. Recovery might be needed, though...
Does this loop fail sometimes but is able to be reestablished?